Automattic/simplenote-macos
# Lists: Undo Exceptions Mark II

### Fix
In this PR we're reimplementing the mechanism(s) that deal with Tabs and Newlines:

- Newlines now insert a **SPTextAttachment directly** when it's needed (fixes UndoManager crashes)
- Tabs are now allowed anywhere within a line that contains (any kind) of bullet: The whole line is indented!
- Unit tests for everything and anything!

@aerych @diegoreymendez gentlemen, may I bug you with a really fun PR?

I've split the test scenarios in two, so that it's not massive for anyone.

**Thank you both in advance!!!**

Ref. #467

---

@aerych Whenever you have a few, mind taking a peek at these? 👇 

### Test: Tabs on Plain Text
1. Add random text on a new note
2. Place the cursor anywhere
3. Press **Tab**

- [x] Verify the plain text is properly indented

### Test: Tabs on Selected Text
1. Add random text on a new note
2. Select Everything
3. Press **Tab**

- [x] Verify the entire text is replaced with a tab!

### Test: Tabs on Lists
1. Insert a list (**Format** > **Insert Checklist**)
2. Enter few items

- [x] Verify that pressing `tab` anywhere on any line indents such line
- [x] Verify that selecting text, and pressing `tab` indents **only** the first line

**Note:** we might revisit the selected-text-indented behavior. Preserving the current `develop` behavior for simplicity.


---

@diegoreymendez Can I bother you with these scenarios? 👇 

### Test: Newline on Empty List Item
1. Insert a list (**Format** > **Insert Checklist**)
2. Hit **newline** on the empty list item

- [x] Verify the list marker is removed

### Test: Cursor Before List Item
1. Insert a list (**Format** > **Insert Checklist**)
2. Place the cursor on the left hand side of the list marker
3. Press newline

- [x] Verify the checklist is moved to the new line
- [x] Verify that we don't end up with two checklist markers on the same line

### Test: Newline on List Item
1. Insert a list (**Format** > **Insert Checklist**)
2. Place the cursor before the list marker, and add few spaces
3. Place the cursor on the right hand side of the marker
4. Insert some random text
5. Hit **newline**

- [x] Verify a new line is added
- [x] Verify that the new line gets the exact same spaces you've added in (2)

---

@diegoreymendez and @aerych Can you both give these a shot please?👇 

1. Add a new note
2. Enter some random title
3. Insert a list (**Format** > **Insert Checklist**)
4. Type some random text for the first item
5. Edit the Title: add a character
6. Click over the list item you've added in (4): hit newline and add new text
6. Repeat steps 4 thru 6 a few times

- [x] Verify that hitting CMD + Z does not break the Undo Manager (no errors in the console!)

**Note:** The Undo Manager seems to be "saving" checkpoints whenever the "last edited spots" aren't contiguous. That's why we need to edit the TITLE > Item (and loop!).

**Thank you gentlemen!!!**

### Release
These changes do not require release notes.
